components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/password/NotificationPasswordRecoveryManager.java

```diff
@@ -249,16 +249,25 @@ private void checkAccountPendingStatus(User user) throws IdentityRecoveryExcepti
     private UserRecoveryData generateNewConfirmationCode(User user, String notificationChannel)
             throws IdentityRecoveryException {
 
+        String recoveryFlowId = null;
         UserRecoveryDataStore userRecoveryDataStore = JDBCRecoveryDataStore.getInstance();
+        UserRecoveryData userRecoveryData = userRecoveryDataStore.loadWithoutCodeExpiryValidation(user);
+        if (userRecoveryData != null) {
+            recoveryFlowId = userRecoveryData.getRecoveryFlowId();
+        }
         userRecoveryDataStore.invalidate(user);
         String secretKey = Utils.generateSecretKey(notificationChannel, user.getTenantDomain(),
                 RecoveryScenarios.NOTIFICATION_BASED_PW_RECOVERY.name());
-        UserRecoveryData recoveryDataDO = new UserRecoveryData(user, secretKey,
+        if (recoveryFlowId != null && !StringUtils.equals(notificationChannel,
+                NotificationChannels.SMS_CHANNEL.getChannelType())) {
```

Comment:
As mentioned above, have this condition the other way. check and correct in all similar places.

Reply:
Fixed with 5e589a260b9c381b606af80b9f02b88198e6830d